### PR TITLE
WEB-1978: Use .pruneKey() to have non-EN JS files use EN files

### PR DESCRIPTION
**Pull request by jeresig**
This uses the new logic added here:
https://github.com/Khan/superagent-cache-plugin/pull/2

To make it so that we can modify the cache key of superagent-cache-plugin, making it so that the URI is now always "en", regardless of the locale. This is safe to do as the filenames are actually unique: they contain unique hashes in them (which we change if they've been translated). If a file hasn't had an translation done to them then the file is essentially the same as the en version so we can use them interchangably, allowing us to cut down on the number of files we need to download and store.

Issue: https://khanacademy.atlassian.net/browse/WEB-1978

Test plan:
I ran `yarn test` and it passed.

### CODE REVIEW

File: src/fetch_package.js [hunk 1]
```diff
@@ -140,6 +140,22 @@ export default async function fetchPackage(
...
+                // have to download far fewer files and use less memory,
+                // especially for our shared libs. This is some
+                // extremely-Khan-specific code and really only works because
+                // all of our JS files have unique hashes embedded in their
+                // names and those hashes change if the contents have been
+                // translated.
+                uri: key.uri.replace(
+                    /\/genwebpack\/prod\/[^/]+\//,
+                    "/genwebpack/prod/en/",
+                ),
```
Comment (kevinbarabash) [inline src/fetch_package.js]: I like that we don't have to the hashes ourselves.

File: src/fetch_package_test.js [hunk 2]
```diff
@@ -179,6 +179,34 @@ describe("fetchPackage with cache", () => {
...
+            "https://www.ka.org/genwebpack/prod/es/ok.js",
+            "TEST",
+        );
+
+        // Assert
+        assert.equal(result0.content, result1.content);
+        // We should still have pending mocks; the second request
+        // should never have gotten sent.
+        assert.notEqual(0, mockScope.pendingMocks().length);
+    });
```
Comment (kevinbarabash) [inline src/fetch_package_test.js]: I'm not familiar with `nock`, but looking at the previous test, this test makes sense to me.

File: src/fetch_package.js [hunk 3]
```diff
@@ -140,6 +140,22 @@ export default async function fetchPackage(
...
+                ...key,
+                // We modify the URI used for the cache key to make it so that
+                // un-translated JS files (that are the same as the en-version)
+                // use the same URI as a cache key. This makes it so that we
+                // have to download far fewer files and use less memory,
+                // especially for our shared libs. This is some
+                // extremely-Khan-specific code and really only works because
+                // all of our JS files have unique hashes embedded in their
+                // names and those hashes change if the contents have been
+                // translated.
```
Comment (somewhatabstract) [inline src/fetch_package.js]: SUPER NIT: In the render server code, I have tried to use block comments for...comment blocks.

File: src/fetch_package.js [hunk 4]
```diff
@@ -140,6 +140,22 @@ export default async function fetchPackage(
         return fetcher
             .use(superagentCache)
             .expiration(24 * 60 * 60)
+            .pruneKey((key) => ({
```
Comment (somewhatabstract) [inline src/fetch_package.js]: This may seem a little extra, but could you split this out into its own function with tests? I think it would be helpful to document the edge cases of what does and does not expect to be modified, just to be sure we're covering what we think we are.

**Comment by somewhatabstract** [comment]
> LGTM
> 
> Some inline suggestions. One I'd really like to see; the other is a super nit of no consequence.

I think that graphql-gateway may use `nock` too. It won't fix the dangling promises issue though. If a test case doesn't handle the settlement of the promises it creates, there will be issues, regardless of sinon or nock or anything else.

File: src/fetch_package.js [hunk 5]
```diff
@@ -80,6 +80,30 @@ export function flushUnusedCache() {
...
+ * JS files (that are the same as the en-version) use the same URI as a cache
+ * key. This makes it so that we have to download far fewer files and use less
+ * memory, especially for our shared libs. This is some extremely-Khan-specific
+ * code and really only works because all of our JS files have unique hashes
+ * embedded in their names and those hashes change if the contents have been
+ * translated.
+ *
+ * @param {object} key A cache key that comes from superagent-cache-plugin
+ */
+export function pruneKey(key: {uri: string, ...}) {
```
Comment (somewhatabstract) [inline src/fetch_package.js]: Great! Thanks for hoisting this into its own export.

File: src/fetch_package_test.js [hunk 6]
```diff
@@ -254,3 +282,41 @@ describe("fetchPackage with cache", () => {
...
+
+        // Act
+        const result = pruneKey(key);
+
+        // Assert
+        assert.deepEqual(result, {
+            other: true,
+            uri: "/genwebpack/prod/en/foo.js",
+        });
+    });
```
Comment (somewhatabstract) [inline src/fetch_package_test.js]: Great! These test cases really help document our intentions. Awesome!

File: package.json [hunk 7]
```diff
@@ -55,7 +55,7 @@
     "node-fetch": "^2.6.0",
     "nodemon": "^1.18.7",
     "superagent": "^5.1.0",
-    "superagent-cache-plugin": "^2.1.0",
+    "superagent-cache-plugin": "git+https://git@github.com/Khan/superagent-cache-plugin.git#803d5170d3e1cd50075acda2eeb6645c8e3b53e1",
```
[the remark this answers is not in the document]
Comment (jeresig) [reply package.json]: Yeah, that'd be really great... sigh...
